main.rkt

```diff
@@ -534,58 +534,18 @@
 
 (define (classify-ival-periodic x period)
   (match-define (ival (endpoint xlo xlo!) (endpoint xhi xhi!) xerr? xerr) x)
-  (define lo* (+ (bigfloat-exponent xlo) (bigfloat-precision xlo)))
-  (define hi* (+ (bigfloat-exponent xhi) (bigfloat-precision xhi)))
-
-  
-  (if (or (and (< 1073741822 (abs lo*)) (not (equal? -9223372036854775807 lo*)))  ; -9223372036854775807 is a code for 0.bf
-          (and (< 1073741822 (abs hi*)) (not (equal? -9223372036854775807 hi*)))) ; |lo* or hi*| > 1073741822 is nan/inf/overflow 
-      'too-wide                     ; overflow/inf/nan
-      (if (<= lo* 0)                ; if lo* belongs to (-1, 1)
-          (if (<= hi* 0)            ; if hi* belongs to (-1, 1)  
-              'near-0
-              (cond
-                [(equal? period 'pi)
-                 (if (>= hi* 2)     ; if hi* belongs to (-inf, -2] U [2, +inf)
-                     'too-wide
-                     'range-reduce)]
-                [(equal? period '2pi)
-                 (if (>= hi* 4)     ; if hi* belongs to (-inf, -8] U [8, +inf)
-                     'too-wide
-                     'range-reduce)]))
-          (if (<= lo* 1)            ; if lo* belongs to (-2, 2)
-              (cond
-                [(equal? period 'pi)
-                 (if (>= hi* 4)     ; if hi* belongs to (-inf, -8] U [8, +inf)
-                     'too-wide
-                     'range-reduce)]
-                [(equal? period '2pi)
-                 (if (>= hi* 5)     ; if hi* belongs to (-inf, -16] U [16, +inf)
-                     'too-wide
-                     'range-reduce)])
-              (if (<= lo* 2)        ; if lo* belongs to (-4, 4)
-                  (cond
-                    [(equal? period 'pi)
-                     (if (>= hi* 4) ; if hi* belongs to (-inf, -8] U [8, +inf)
-                         'too-wide
-                         'range-reduce)]
-                    [(equal? period '2pi)
-                     (if (>= hi* 5) ; if hi* belongs to (-inf, -16] U [16, +inf)
-                         'too-wide
-                         'range-reduce)])
-                  (if (>= lo* 3)    ; if lo* belongs to (-inf, -4] U [4, +inf)
-                      (if (>= hi* 3) ; if hi* belongs to (-inf, -4] U [4, +inf)
-                          (if (> (abs (- hi* lo*)) 1)
-                              'too-wide
-                              (if (equal? (bigfloat-signbit xlo) (bigfloat-signbit xhi))
-                                  'range-reduce
-                                  'too-wide))
-                          'range-reduce)
-                      'range-reduce))))))
+  (define lo-ulp (bigfloat-exponent xlo))
+  (define hi-ulp (bigfloat-exponent xhi))
+  (define lo-exp (+ lo-ulp (bigfloat-precision xlo)))
+  (define hi-exp (+ hi-ulp (bigfloat-precision xhi)))
+  (cond
+    [(and (not (bfinfinite? xlo)) (not (bfinfinite? xhi))
+          (< lo-exp 0) (< hi-exp 0)) 'near-0]
```

Comment:
We may have to add `(bfnan?)` conditions, because exponent of `+nan.bf` is `-2147483646` - which can be considered as `'near-0` case. Another value I am thinking of is: `min.bf = (bf "-2.382564904887951073216169781732674520415e-323228497")`, the exponent is equal to `-1073741823`.
We maybe want check whether `lo-exp < 1` = `xlo < 1.bf`, depends whether we want to classify values closer to `1.bf` as `'near-0` as well.

Reply:
Hmm, ok. I don't know if we need to deal with `+nan.bf`. It shouldn't be possible to have `+nan.bf` as the endpoint of an interval unless it's the illegal interval, in which case we can special-case that.

I think min.bf should be handled fine, right? It's exponent is small, so it'll be `near-0`, which I think is fine?

Comparing the exp to 1 instead of 0 is fine but would require reasoning about the periodicity. I imagine it doesn't actually matter that much: since these values are small, the precision we ultimately choose for the range reduction will also be small, and it's only one binade.

Reply:
Yeah, I see, `min.bf` is not a problem. 
By the way, that's impressive how this implementation fits Herbie